File: plugins/agentic-sdlc/src/agentic_sdlc/commands/status.py
"""Status, cancel, and list command handlers."""

from __future__ import annotations

import json
import sys
from pathlib import Path
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from argparse import Namespace


def cmd_status(args: Namespace) -> None:
    """Show workflow status."""
    from agentic_sdlc.progress import load_progress

    progress = load_progress(args.workflow_id)
    if progress is None:
        print(f"Error: Workflow not found: {args.workflow_id}", file=sys.stderr)
        sys.exit(1)

    print(f"Workflow: {progress.workflow_name}")
    print(f"ID: {progress.workflow_id}")
    print(f"Status: {progress.status}")
    print(f"Started: {progress.started_at}")
    if progress.completed_at:
        print(f"Completed: {progress.completed_at}")

    if progress.current_step:
        print(f"\nCurrent Step: {progress.current_step['name']}")
        print(f"  Retry Count: {progress.current_step.get('retry_count', 0)}")

    if progress.completed_steps:
        print("\nCompleted Steps:")
        for step in progress.completed_steps:
            status_icon = "+" if step.status == "completed" else "x"
            print(f"  [{status_icon}] {step.name}")

    if progress.pending_steps:
        print("\nPending Steps:")
        for step_name in progress.pending_steps:
            print(f"  [ ] {step_name}")

    if progress.errors:
        print("\nErrors:")
        for error in progress.errors:
            print(f"  - {error['step']}: {error['error']}")


def cmd_cancel(args: Namespace) -> None:
    """Cancel a running workflow."""
    from datetime import datetime, timezone

    from agentic_sdlc.progress import WorkflowStatus, load_progress, save_progress

    progress = load_progress(args.workflow_id)
    if progress is None:
        print(f"Error: Workflow not found: {args.workflow_id}", file=sys.stderr)
        sys.exit(1)

    if progress.status not in [WorkflowStatus.RUNNING.value, WorkflowStatus.PAUSED.value]:
        print(
            f"Error: Cannot cancel workflow in '{progress.status}' status",
            file=sys.stderr,
        )
        sys.exit(1)

    progress.status = WorkflowStatus.CANCELLED.value
    progress.completed_at = datetime.now(timezone.utc).isoformat()
    save_progress(progress)

    print(f"Workflow cancelled: {args.workflow_id}")


def cmd_list(args: Namespace) -> None:
    """List workflows."""
    workflows_dir = Path.cwd() / "agentic" / "workflows"
    if not workflows_dir.exists():
        print("No workflows found.")
        return

    workflows = []
    for workflow_dir in workflows_dir.iterdir():
        if workflow_dir.is_dir():
            progress_file = workflow_dir / "progress.json"
            if progress_file.exists():
                with open(progress_file) as f:
                    data = json.load(f)
                    if args.status is None or data.get("status") == args.status:
                        workflows.append(data)

    if not workflows:
        print("No workflows found." + (f" (status={args.status})" if args.status else ""))
        return

    print(f"{'ID':<12} {'Name':<25} {'Status':<12} {'Started':<20}")
    print("-" * 70)
    for wf in workflows:
        started = wf.get("started_at", "")[:19] if wf.get("started_at") else ""
        print(f"{wf.get('workflow_id', ''):<12} {wf.get('workflow_name', '')[:25]:<25} {wf.get('status', ''):<12} {started:<20}")

File: src/commands/status.ts
/** Status, cancel, and list command handlers. */

import { existsSync, readFileSync, readdirSync } from "node:fs";
import path from "node:path";

import { WORKFLOW_STATUS, loadProgress, saveProgress } from "../progress.js";

export function cmdStatus(workflowId: string): void {
	const progress = loadProgress(workflowId);
	if (progress === null) {
		process.stderr.write(`Error: Workflow not found: ${workflowId}\n`);
		process.exit(1);
	}

	process.stdout.write(`Workflow: ${progress.workflowName}\n`);
	process.stdout.write(`ID: ${progress.workflowId}\n`);
	process.stdout.write(`Status: ${progress.status}\n`);
	process.stdout.write(`Started: ${progress.startedAt}\n`);
	if (progress.completedAt) {
		process.stdout.write(`Completed: ${progress.completedAt}\n`);
	}

	if (progress.currentStep) {
		process.stdout.write(`\nCurrent Step: ${progress.currentStep.name}\n`);
		process.stdout.write(`  Retry Count: ${progress.currentStep.retryCount}\n`);
	}

	if (progress.completedSteps && progress.completedSteps.length > 0) {
		process.stdout.write("\nCompleted Steps:\n");
		for (const step of progress.completedSteps) {
			const icon = step.status === "completed" ? "+" : "x";
			process.stdout.write(`  [${icon}] ${step.name}\n`);
		}
	}

	if (progress.pendingSteps && progress.pendingSteps.length > 0) {
		process.stdout.write("\nPending Steps:\n");
		for (const stepName of progress.pendingSteps) {
			process.stdout.write(`  [ ] ${stepName}\n`);
		}
	}

	if (progress.errors && progress.errors.length > 0) {
		process.stdout.write("\nErrors:\n");
		for (const error of progress.errors) {
			const err = error as Record<string, string>;
			process.stdout.write(`  - ${err.step}: ${err.error}\n`);
		}
	}
}

export function cmdCancel(workflowId: string): void {
	const progress = loadProgress(workflowId);
	if (progress === null) {
		process.stderr.write(`Error: Workflow not found: ${workflowId}\n`);
		process.exit(1);
	}

	if (progress.status !== WORKFLOW_STATUS.RUNNING && progress.status !== WORKFLOW_STATUS.PAUSED) {
		process.stderr.write(`Error: Cannot cancel workflow in '${progress.status}' status\n`);
		process.exit(1);
	}

	progress.status = WORKFLOW_STATUS.CANCELED;
	progress.completedAt = new Date().toISOString();
	saveProgress(progress);

	process.stdout.write(`Workflow canceled: ${workflowId}\n`);
}

export function cmdList(statusFilter?: string): void {
	const outputsDir = path.join(process.cwd(), "agentic", "outputs");
	if (!existsSync(outputsDir)) {
		process.stdout.write("No workflows found.\n");
		return;
	}

	const workflows: Record<string, unknown>[] = [];
	const entries = readdirSync(outputsDir, { withFileTypes: true });
	for (const entry of entries) {
		if (entry.isDirectory()) {
			const progressFile = path.join(outputsDir, entry.name, "progress.json");
			if (existsSync(progressFile)) {
				const data = JSON.parse(readFileSync(progressFile, "utf-8")) as Record<string, unknown>;
				if (statusFilter == null || data.status === statusFilter) {
					workflows.push(data);
				}
			}
		}
	}

	if (workflows.length === 0) {
		const suffix = statusFilter ? ` (status=${statusFilter})` : "";
		process.stdout.write(`No workflows found.${suffix}\n`);
		return;
	}

	process.stdout.write(
		`${"ID".padEnd(12)} ${"Name".padEnd(25)} ${"Status".padEnd(12)} ${"Started".padEnd(20)}\n`,
	);
	process.stdout.write(`${"-".repeat(70)}\n`);
	for (const wf of workflows) {
		const id = String(wf.workflow_id ?? "").padEnd(12);
		const name = String(wf.workflow_name ?? "")
			.slice(0, 25)
			.padEnd(25);
		const status = String(wf.status ?? "").padEnd(12);
		const started = wf.started_at ? String(wf.started_at).slice(0, 19).padEnd(20) : "".padEnd(20);
		process.stdout.write(`${id} ${name} ${status} ${started}\n`);
	}
}
